### ISSUE TITLE
Add a rate-limiting mechanism for pelias queries, make them cache-friendlier and don't bother sending requests with 1 or 2 latin characters only.

### ISSUE THREAD

**Issue by ellenhp** [headwaymaps/headway#118]
### Rate-limiting

The way I imagine this is a caching layer on top of pelias requests that takes updates to the search box and computes the Levenshtein distance of the query to everything we have in the cache, immediately switching the features list if something in the cache is closer in edit distance than what's currently displayed.

Additionally, each time the search box contents changes run through the following:

1. If the edit distance to the best-known results is more than two, send a request immediately
2. if the edit distance is two, send a request no sooner than 150ms after the previous one (unless edit distance changes).
3. if the edit distance is one, send a request no sooner than 300ms after the previous one (unless edit distance changes).

What I'm going for here is two-fold:

- Stale data should not be displayed unless it's truly the best we have (protect against more recent query results being overwritten by a slower, less recent query)
- Fast typists shouldn't necessarily get the same number of requests sent out during a query as slow typists. Autocomplete can really help a slow typist so we should definitely use it, but for a fast typist (maybe on a slow network) who's done typing "starbucks" faster than 2x or 3x their ping time to Hetzner, we should probably just send one after `sta`, `starb` and `starbucks`

We'll see how that feels, it might add too much jank to have those timeouts explicitly coded in there, but I think it'll be fine.

### Cache-friendlier pelias requests

Viewport biasing points should be quantized to a grid so that small changes in viewport don't cachebust needlessly. Viewport biasing in Pelias is very weak right now so this should be fine.

### Don't bother sending short requests

I'm only comfortable doing this for `[a-zA-Z]` but it could help ensure that the first batch of results returned is more helpful.

**Comment by ellenhp** [comment]
>The way I imagine this is a caching layer on top of pelias requests that takes updates to the search box and computes the Levenshtein distance of the query to everything we have in the cache, immediately switching the features list if something in the cache is closer in edit distance than what's currently displayed.

The more I think about it, the more this sounds like it's actually a bad idea. Edit distance isn't a good approximation of semantic similarity. We should definitely cache pelias responses though, but I think we should only use them if they're exact matches. If we wanted to get fancy we could also cache prefixes. Probably not really necessary though because assuming the characters came in sequentially we already kind of do that.

**Comment by michaelkirk** [comment]
> We should definitely cache pelias responses though, but I think we should only use them if they're exact matches. 

If I'm understanding correctly, this already appears to be working for maps.earth.

On maps.earth, the http cache appears to already be caching requests (maybe related: https://github.com/pelias/api/issues/23). Note the "**Transfer**: `(disk)`" for `autocomplete` requests in the network inspector:

https://user-images.githubusercontent.com/217057/191567053-37a09ad9-f7c3-4d36-8390-6679c93a7767.mp4

Interestingly, when I run the frontend on `quasar dev` all my queries have a different etag. I wouldn't be surprised if `quasar dev` has some intentional cache-busting mechanism, but I haven't dug in enough to verify if that's the case. Presumably that's fine/desirable behavior for dev:

https://user-images.githubusercontent.com/217057/191567094-52a15cb0-4a69-48a4-a5d9-656d39053a8d.mp4

**Comment by michaelkirk** [comment]
> Don't bother sending short requests

All the mapping products I've tried will autocomplete even single letter requests. It's possible to get somewhat reasonable results, especially if you're zoomed in a bit.

I think I understand the rationale of the proposal, but my instinct is that it's not worth going to the effort to disable something which is commonly done, makes the control's behavior *slightly* less explicable, and *might* actually return useful results.

In summary: I think it's reasonable to consider this issue as 
"Done/WONTFIX".

What do you think?

**Comment by michaelkirk** [comment]
Oh yeah whoops, throttling is still a good idea.
